akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -68,19 +70,37 @@ private[remote] object EnvelopeBuffer {
 
 /**
  * INTERNAL API
+ * Decompress and cause compression advertisements.
  */
-private[remote] trait LiteralCompressionTable {
+private[remote] trait InboundCompression {
+  def hitActorRef(remote: Address, ref: ActorRef): Unit
+  def decompressActorRef(idx: Int): OptionVal[ActorRef]
 
-  def compressActorRef(ref: String): Int
-  def decompressActorRef(idx: Int): String
+  def hitClassManifest(remote: Address, manifest: String): Unit
+  def decompressClassManifest(idx: Int): OptionVal[String]
+}
+/**
+ * INTERNAL API
+ * Compress outgoing data and handle compression advertisements to fill compression table.
+ */
+private[remote] trait OutboundCompression {
+  def allocateActorRefCompressionId(ref: ActorRef, id: Int): Unit
+  def compressActorRef(ref: ActorRef): Int
 
+  def allocateClassManifestCompressionId(manifest: String, id: Int): Unit
   def compressClassManifest(manifest: String): Int
-  def decompressClassManifest(idx: Int): String
-
 }
 
 object HeaderBuilder {
-  def apply(compressionTable: LiteralCompressionTable): HeaderBuilder = new HeaderBuilderImpl(compressionTable)
+
+  // We really only use the Header builder on one "side" or the other, thus in order to avoid having to split its impl
+  // we inject no-op compression's of the "other side".  
+
+  def in(compression: InboundCompression): HeaderBuilder = new HeaderBuilderImpl(compression, NoopOutboundCompression)
+  def out(compression: OutboundCompression): HeaderBuilder = new HeaderBuilderImpl(NoopInboundCompression, compression)
+
+  /** INTERNAL API, FOR TESTING ONLY */
+  private[remote] def bothWays(in: InboundCompression, out: OutboundCompression): HeaderBuilder = new HeaderBuilderImpl(in, out)
```

Comment:
Not happy about exposing it here, but saves us a lot of dancing around in `EnvelopeBufferSpec` (which both reads/writes the same one)